nm: Increase minimum supported NetworkManager version to 1.40

cathay4t opened the pull request:
Increasing minimum supported NetworkManager version to 1.40 where MPTCP
and `802-3-ethernet.accept-all-mac-addresses` property are supported.

With this, we could discard unnecessary call to NetworkManager daemon
for checking support status.

Nmstate only generate warning message for unsupported NM version.

cathay4t commented:
Background:
 * Fedora 38 ships NM 1.42
 * RHEL 8.8 ships NM 1.40
 * RHEL 9.2 ships NM 1.42
 * Debian stable ships NM 1.42
 * Ubuntu 22.03 LTS ships NM 1.36 (Ubuntu will release LTS in 2024)

cathay4t commented:
The reason I do this is because I noticed we invoke `nm_api.version()` every time we create a nm wire setting, instead of patching it to better workflow, I think it is time to bump NM versions.

ihuguet commented:
I'm not sure about this change.
1. You still call to NM daemon to check if the version is >= 1.40 or not, so there is no real gain regarding number of calls to NM.
2. Users can get confused if they see a Warning of NM < 1.40 unsupported. Actually only MPTCP is unsupported, but maybe the user is not using MPTCP.
3. Version 1.36 is still widely used.

Would it be possible to do the version check only if we're trying to apply MPTCP options? Then, if the user is trying to apply MPTCP options and the version is <1.40, reject to apply. Or emit a warning and let NM to reject the settings. This way we would also save calls to NM, checking the version only if necessary.

liangwen12year commented:
Since we are bumping the minimum supported version for NM, I also saw the updated code removing the conditional statement on support status of mptcp flag, can we also remove the corresponding version check on the integration tests for determining if the test should be skipped ? 

for example: 
```
@pytest.mark.skipif(
    nm_minor_version() < 40, reason="MPTCP is not supported by NetworkManager"
)
```

cathay4t commented:
> I'm not sure about this change.
> 
>     1. You still call to NM daemon to check if the version is >= 1.40 or not, so there is no real gain regarding number of calls to NM.

It will be a performance penalty when creating 1000+ veth.

> 
>     2. Users can get confused if they see a Warning of NM < 1.40 unsupported. Actually only MPTCP is unsupported, but maybe the user is not using MPTCP.
> 
>     3. Version 1.36 is still widely used.

Widely used does not mean we support that use environment. Can you be specific on which release?

I only see Ubuntu latest LTS still use 1.3x NM, considering they are release 2024 LTS this year, I think it is acceptable to drop support of NM 1.40<

> 
> 
> Would it be possible to do the version check only if we're trying to apply MPTCP options? Then, if the user is trying to apply MPTCP options and the version is <1.40, reject to apply. Or emit a warning and let NM to reject the settings. This way we would also save calls to NM, checking the version only if necessary.

Technically can do. Besides what you mentioned, we also put extra effort to make sure our CI covered this old NM version and review any failure caused by it.

So the question is whether it is worth it to maintain code for use environment we do not support or even just from our imagination?

cathay4t commented:
> Since we are bumping the minimum supported version for NM, I also saw the updated code removing the conditional statement on support status of mptcp flag, can we also remove the corresponding version check on the integration tests for determining if the test should be skipped ?
> 
> for example:
> 
> ```
> @pytest.mark.skipif(
>     nm_minor_version() < 40, reason="MPTCP is not supported by NetworkManager"
> )
> ```

Nice catch! Will do.

ihuguet commented:
Ubuntu 22.04 is supported until April 2027, so it is clearly not only on our imagination. RHEL-9.0 also uses 1.36 and it has EUS support until June and SAP support 4 years more.

OK, they are not shipping latest nmstate so we can just drop the support. But both solutions are equally easy to implement and with mine we would save an unnecessary call to NM in most situations (I agree that the performance lost is negligible, though).

Anyway, take it as an improvement suggestion, it is OK as it is now, too.

liangwen12year commented:
I still found some tests containing the skip condition `nm_major_minor_version() < 1.31`, the following can also be removed ? 

```
diff --git a/tests/integration/interface_common_test.py b/tests/integration/interface_common_test.py
index 6057fca7..65903f1b 100644
--- a/tests/integration/interface_common_test.py
+++ b/tests/integration/interface_common_test.py
@@ -124,10 +124,6 @@ def test_take_over_virtual_interface_and_rollback(ip_link_dummy):
         assert len(current_state[Interface.KEY]) == 1
 
 
-@pytest.mark.skipif(
-    nm_major_minor_version() < 1.31,
-    reason="Modifying accept-all-mac-addresses is not supported on NM.",
-)
 def test_enable_and_disable_accept_all_mac_addresses(eth1_up):
     desired_state = eth1_up
     desired_state[Interface.KEY][0][Interface.ACCEPT_ALL_MAC_ADDRESSES] = True
diff --git a/tests/integration/linux_bridge_test.py b/tests/integration/linux_bridge_test.py
index ffc586c3..82cc8818 100644
--- a/tests/integration/linux_bridge_test.py
+++ b/tests/integration/linux_bridge_test.py
@@ -968,11 +968,6 @@ def test_create_linux_bridge_with_copy_mac_from_permanent_mac(
         assert_mac_address(current_state, permanent_mac)
 
 
-@pytest.mark.tier1
-@pytest.mark.skipif(
-    nm_major_minor_version() < 1.31,
-    reason="Modifying accept-all-mac-addresses is not supported on NM.",
-)
 def test_linux_bridge_enable_and_disable_accept_all_mac_addresses(
     bridge0_with_port0,
 ):
```

cathay4t committed:
nm: Increase minimum supported NetworkManager version to 1.40

Increasing minimum supported NetworkManager version to 1.40 where MPTCP
and `802-3-ethernet.accept-all-mac-addresses` property are supported.

With this, we could discard unnecessary call to NetworkManager daemon
for checking support status.

Nmstate only generate warning message for unsupported NM version.

Removed un-required version check in integration test cases.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
> I still found some tests containing the skip condition `nm_major_minor_version() < 1.31`, the following can also be removed ?

Removed. Thanks!

liangwen12year commented:
Please rebase to resolve the conflicts.

ffmancera reviewed (approved):
I am fine with the change, as this is basically affecting main branch. I do not expect old releases using NM 1.3X will ship latest Nmstate version.
